### ISSUE TITLE
Add Year & Month types

### ISSUE THREAD

**Issue by Neme12** [nodatime/nodatime#1468]
I would find it useful if NodaTime included additional `Year` and `Month` data types (`Month` not being the same as `YearMonth` - `YearMonth` would be a combination of these two).

**Why? Isn't that just a number?**
Yes, storing a year or a month as an integer is totally fine. But such a type would still be useful for formatting.

Today, if I want to format a year and month in a localized way (to get something like "January 2020"), I have to create a `LocalDate` with the desired year and a month, and a placeholder day of month. This is where the `YearMonth` data type comes handy. To me, this is not only a simplification of the code to avoid having to use placeholder/dummy values for a date. It's important that we only store the data that we actually know and have a proper data type for that which would only allow us to do operations that make sense there - this is as far as I can tell one of the core principles of NodaTime. Storing year & month in a `LocalDateTime` would to me seem like using `System.DateTime` to only represent date values without time.

The same can be said for month-only values. What if I want to store a month and then format it in a localized way - to get "January"? It seems I'd have to create a `YearMonth` with a dummy year. With years, this would admittedly be a somewhat less useful because the only localized part of a formatted year is usually the era - which most people don't deal with, but it would still provide value and in my opinion would be nice for consistency.

**Additional benefits**

Having these types would make some other operations a bit easier (or at the very least more discoverable) as we wouldn't have to look for helper functions in `CalendarSystem` and elsewhere. A `Month` type could provide a `DaysInMonth` property. A `Year` type could provide `MonthsInYear`, `DaysInYear` and `IsLeapYear` properties. Operations such as getting the last day in a year + month, or getting the last day of a year would  be a lot more straightforward.
Instead of
```c#
// last day of year
new LocalDate(year, CalendarSystem.Iso.GetMonthsInYear(year), CalendarSystem.Iso.GetDaysInMonth(CalendarSystem.Iso.GetMonthsInYear(year)))
```
we could have something like
```c#
new Year(year).AtLastMonth().AtEndOfMonth()
// or
new Year(year).AtEndOfYear()
```

A `Year` data type would be constructible from either an absolute year or a year + era, and would have properties for both. This would allow us to reduce duplication as these operations are currently  scattered in a lot of places - but this would require a huge break.
Instead of data types like  `YearMonth`, `LocalDate`, `LocalDateTime` etc having 3 properties - `Year`, `YearOfEra` and `Era`, they would only have one - `Year`, which would return a `Year` data type which would provide these 3 individual properties.

**Comment by jskeet** [comment]
I'll think about it - but each type brings extra cognitive load, as well as all the extra actual implementation. Given that newcomers are already often overwhelmed by the number of types in NodaTime, adding an extra two (or three - it would be a little odd to have Month and Year but not Day) increases this.

**Comment by jskeet** [comment]
(I definitely wouldn't want to change the type of the normal Year and Month properties. I really don't think this is worth breaking the code of all existing users, even though for 3.0 *could* theoretically break as much as we want.)

**Comment by Neme12** [comment]
> or three - it would be a little odd to have Month and Year but not Day

That's interesting, I never thought about that one. I'm used to using the [js-joda](https://js-joda.github.io/js-joda/) library in JavaScript, which has `Year` and `Month` - but no `Day` type, and I miss the former 2 (and also `YearMonth`, since it's not in a stable version yet) when I switch to C# and NodaTime. It seems the library is inspired by java.time, which, looking at the [API documentation](https://docs.oracle.com/javase/8/docs/api/java/time/package-summary.html), also seems to have `Year` and `Month` types but no `Day` type.

I can't imagine a `Day` type being useful. Is there a calendar system/culture where the formatted day of month isn't always just an integer formatted using the same culture?

I especially like having `Year` + `YearMonth` because it gives me the ability to combine them into a more specific type - for example `year.AtMonth(month)` (returns `YearMonth`) and `yearMonth.AtDay(day)` (returns `LocalDate`). `Month` could be combined with a day of month to get a `MonthDay` (called `AnnualDate` in NodaTime), but I couldn't do anything with `Day` on its own, which also makes me think there's no value in having it.

Another worry is that the maximum allowed day of month depends on the month, so it would seem a little weird to have a day of month without a month (although I'm guessing there are calendar systems where the maximum month depends on the year? If so then the same argument could be used against having a `Month` type).

**Comment by Neme12** [comment]
I use these types a lot in client applications with calendar-like components and generally where the UI allows the user to navigate between years/months and show different data for each, or generally for formatting years and month names.

This is not an uncommon use case but I think that most libraries/applications out there just use their own localization for month names as part of the app, which I think is unnecessary and unfortunate because in case of a full date, they do use the system localization (which of course includes month names), which creates an inconsistency.

**Comment by Neme12** [comment]
> Given that newcomers are already often overwhelmed by the number of types in NodaTime, adding an extra two (or three - it would be a little odd to have Month and Year but not Day) increases this.

That's true, but I think that users are mostly overwhelmed with choosing the Local/Offset/Zoned date-time variants and Instant. Types like `Year` and `Month` make it perfectly clear as to what they represent just from their name and I don't think their presence would lead to more confusion or make it harder for users to figure out what type they need. I think the only confusion might come from users wondering why they would need a `Year` type at all as opposed to using an integer.

**Comment by jskeet** [comment]
I suspect even "Year" and "Month" might be confusing to start with, just in terms of them being available at all... and then even more so when we add calendar systems into the mix.

Note that in some calendars, a month may be quite odd - for example, the Hebrew calendar system with its leap month options means that one month number doesn't always mean the same thing - it depends on the year in which it's in.

I'm not saying "no" to all of this, but I think it'll need quite a lot of design work to make sure it all hangs together. I don't want to block 3.0 on this (or do a half job for 3.0). I'll add a milestone of 3.1-consider.

**Comment by Neme12** [comment]
> Note that in some calendars, a month may be quite odd - for example, the Hebrew calendar system with its leap month options means that one month number doesn't always mean the same thing - it depends on the year in which it's in.

That's definitely something to think about. But from what I understand, isn't there already a data type in NodaTime that represents a month (and a day) without a year - `AnnualDate`?

**Comment by jskeet** [comment]
> That's definitely something to think about. But from what I understand, isn't there already a data type in NodaTime that represents a month (and a day) without a year - `AnnualDate`?

Yes, and that's restricted to the ISO calendar for exactly that reason. I would *prefer* to avoid that restriction if possible. We should think carefully before committing to anything (which is why I've targeted this at 3.1).

**Comment by weitzhandler** [comment]
Related: #1503.
